Lib/posixpath.py

```diff
@@ -412,49 +412,52 @@ def realpath(filename, *, strict=False):
     """Return the canonical path of the specified filename, eliminating any
 symbolic links encountered in the path."""
     filename = os.fspath(filename)
-    path, ok = _joinrealpath(filename[:0], filename, strict, {})
-    return abspath(path)
-
-# Join two paths, normalizing and eliminating any symbolic links
-# encountered in the second path.
-def _joinrealpath(path, rest, strict, seen):
-    if isinstance(path, bytes):
+    if isinstance(filename, bytes):
         sep = b'/'
         curdir = b'.'
         pardir = b'..'
+        getcwd = os.getcwdb
     else:
         sep = '/'
         curdir = '.'
         pardir = '..'
+        getcwd = os.getcwd
+
+    # The stack of unresolved path parts.
+    rest = filename.split(sep)[::-1]
 
-    if isabs(rest):
-        rest = rest[1:]
-        path = sep
+    # The resolved path, which is absolute throughout this function.
+    # Note: getcwd() returns a normalized and symlink-free path.
```

Comment:
On Linux until a relatively recent glibc 2.27 (2018) the string "(unreachable)" could be returned instead of raising an error. https://man7.org/linux/man-pages/man3/getcwd.3.html BUGS section. We should check that the getcwd return value starts with sep before using it to avoid propagating a platform bug.  That can be done in Modules/posixmodule.c on the getcwd() implementation itself.

In case other posix libc+kernel platforms combos could also have their own known warts along these lines it is worth at least checking Darwin, FreeBSD, AIX, and Solaris man pages.

Reply:
Hm! But the current implementation of `realpath()` calls `abspath()`, which can call `getcwd()`, and so I think we're already exposed to that bug, right?

Reply:
We are, but it's raised at a different point. With this PR, when `getcwd` returns `(unreachable)`, then `../..` resolves to... `(unreachabl`! (`rfind('/')` returns -1 so `'..'` chops off a letter).
The existing behaviour isn't perfect either, of course. But it seems like this is worth a check for `/`.

FWIW, `getcwd` man pages for the systems Greg mentioned don't contain notes about warts (local macOS 14, [FreeBSD](https://man.freebsd.org/cgi/man.cgi?query=getcwd&sektion=3&n=1), [AIX](https://www.ibm.com/docs/en/aix/7.3?topic=g-getcwd-subroutine), [Solaris](https://docs.oracle.com/cd/E26502_01/html/E29034/getcwd-3c.html#scrolltoc)).

Reply:
Thanks both. I've put a very tentative fix up as #117481, but my knowledge of C and the the Python C API is pretty minimal, so it might be totally wrong.

Reply:
#117481 is merged, so I'll resolve this comment thread.